deleted references of self.vocab_size and self.type_vocab_size for multiple models [TF implementation]

susnato opened the pull request:
# What does this PR do?

This PR deletes references of  `self.vocab_size` and `self.type_vocab_size` for these models [Tensorflow implementation] : bert, albert, lxmert, electra, tapas, convbert, layoutlm, gpt2, camembert, clip, ctrl, deberta, deberta_v2, distilbert, esm, funnel, gptj, groupvit, longformer, mobilebert, mpnet, openai, rembert, roberta, roberta_prelayernorm, roformer, xlm_roberta, xlnet.
<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->



## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.([link](https://github.com/huggingface/transformers/issues/21053))
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

@gante

susnato commented:
Hi, @gante I made changes for `src/transformers/models/albert/modeling_tf_albert.py`, could you please check it? If it's ok then I will push other changes of other models.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_21164). All of your documentation changes will be reflected on that endpoint.

gante commented on src/transformers/models/albert/modeling_tf_albert.py:
```diff
@@ -138,7 +138,7 @@ class TFAlbertEmbeddings(tf.keras.layers.Layer):
 
-        self.vocab_size = config.vocab_size
+        self.config = config
         self.type_vocab_size = config.type_vocab_size
```
if you have the patience to do it, removing the other attributes that are set from config (like `self.type_vocab_size`) would also be great :) That way there will be a greater similarity between PT and TF 🧡

susnato replied:
@gante 
Ok, I will remove `self.type_vocab_size`, btw there seem to be 3 more lines 
```
self.embedding_size = config.embedding_size
self.max_position_embeddings = config.max_position_embeddings
self.initializer_range = config.initializer_range
```
I don't think we need to change them for this purpose, right?

gante replied:
It's not needed, but it would be a nice refactor :) Don't worry about it if you are not interested in making the change, we can pick it up afterward -- the `vocab_size` is the most important one for now, as it is causing bugs.

gante reviewed (commented):
@susnato LGTM, this is precisely what I wanted to do!

gante commented:
@susnato before moving forward, let's make sure our `transformers` master agrees with the set of changes we are about to make :)

_________________________________
@sgugger this PR contains a change I want to apply across TF models, and @susnato is kindly doing the bulk of the work. In essence, some `config` attributes are not constant throughout the model's life, like `config.vocab_size`, and our TF implementation stores them as immutable class attributes (e.g. `self.vocab_size = config.vocab_size`). PT doesn't have this problem, since it simply stores `self.config = config` in the layers, which benefits from the updates the mutable `config` dictionary may receive elsewhere.

The proposed change is to make TF implementation closer to PT implementation and store `self.config` in the layers, as opposed to individual configuration parameters. This also solves the bug that triggered this discussion, where the vocabulary size was not being correctly updated and causing exceptions.

Let us know if you are okay with us making this change over most model architectures 🚀

sgugger reviewed (approved):
I have always wondered why TensorFlow models did not store the config. Just pinging @Rocketknight1 to also have a look in case there is some TF-arcane behing this that would break something but otherwise all for this change!

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for bert

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for albert

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for lxmert

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for convbert

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for electra

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for layoutlm

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for tapas

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for gpt2

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for camembert

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for clip

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for ctrl

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for deberta

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for deberta_v2

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for distilbert

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for esm

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for funnel

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for gptj

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for groupvit

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for longformer

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for mobilebert

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for mpnet

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for openai

gante commented:
@susnato you might need to run `make fixup` locally, to automatically format the code and make our CI happy

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for rembert

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for roberta

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for roberta_prelayernorm

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for roformer

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for xlm_roberta

susnato committed:
deleted references of self.vocab_size & self.type_vocab_size for xlnet

susnato commented:
Hi, @gante I added all the models I found to have self.vocab_size and removed reference to self.vocab_size and self.type_vocab_size. Could you please check it?

gante reviewed (approved):
LGTM 👍 

Can we edit the PR title to a shorter one before merging? 😅

susnato commented:
> LGTM +1
> 
> Can we edit the PR title to a shorter one before merging? sweat_smile

@gante Done!

gante commented:
Awesome!  

Thank you for all the work you've put into fixing this, @susnato 🤗
